src/SelfServe/SelfServeComponent.tsx

```diff
@@ -173,46 +261,155 @@ export class SelfServeComponent extends React.Component<SelfServeComponentProps,
+  };
+
+  public onRefreshClicked = async (): Promise<void> => {
+    this.setState({ isInitializing: true });
+    const refreshResult = await this.performRefresh();
+    if (!refreshResult.isComponentUpdating) {
+      this.initializeSmartUiComponent();
+    }
+    this.setState({ isInitializing: false });
+  };
+
+  private getCommandBarItems = (): ICommandBarItemProps[] => {
+    return [
+      {
+        key: "save",
+        text: "Save",
+        iconProps: { iconName: "Save" },
+        split: true,
+        disabled: this.isSaveButtonDisabled(),
+        onClick: this.onSaveButtonClick,
+      },
+      {
+        key: "discard",
+        text: "Discard",
+        iconProps: { iconName: "Undo" },
+        split: true,
+        disabled: this.isDiscardButtonDisabled(),
+        onClick: () => {
+          this.discard();
+        },
+      },
+      {
+        key: "refresh",
+        text: "Refresh",
+        disabled: this.state.isInitializing,
+        iconProps: { iconName: "Refresh" },
+        split: true,
+        onClick: () => {
+          this.onRefreshClicked();
+        },
+      },
+    ];
+  };
+
   public render(): JSX.Element {
-    const containerStackTokens: IStackTokens = { childrenGap: 20 };
-    return !this.state.isRefreshing ? (
+    const containerStackTokens: IStackTokens = { childrenGap: 5 };
+    if (this.state.compileErrorMessage) {
+      return <MessageBar messageBarType={MessageBarType.error}>{this.state.compileErrorMessage}</MessageBar>;
+    }
+    return (
       <div style={{ overflowX: "auto" }}>
-        <Stack tokens={containerStackTokens} styles={{ root: { width: 400, padding: 10 } }}>
-          <SmartUiComponent
-            descriptor={this.state.root as SmartUiDescriptor}
-            currentValues={this.state.currentValues}
-            onInputChange={this.onInputChange}
-          />
-
-          <Stack horizontal tokens={{ childrenGap: 10 }}>
-            <PrimaryButton
-              id="submitButton"
-              styles={{ root: { width: 100 } }}
-              text="submit"
-              onClick={async () => {
-                await this.props.descriptor.onSubmit(this.state.currentValues);
-                this.setDefaults();
-              }}
-            />
-            <PrimaryButton
-              id="discardButton"
-              styles={{ root: { width: 100 } }}
-              text="discard"
-              onClick={() => this.discard()}
+        <Stack tokens={containerStackTokens} styles={{ root: { padding: 10 } }}>
+          <CommandBar styles={{ root: { paddingLeft: 0 } }} items={this.getCommandBarItems()} />
+          {this.state.isInitializing ? (
+            <Spinner
+              size={SpinnerSize.large}
+              styles={{ root: { textAlign: "center", justifyContent: "center", width: "100%", height: "100%" } }}
             />
-          </Stack>
+          ) : (
+            <>
+              {this.state.refreshResult?.isComponentUpdating && (
```

Comment:
>isComponentUpdating [](start = 41, length = 19)

updateInProgress.. it is not component updating but actually the rp call to update is still happening